bindings/matrix-sdk-ffi/src/room_directory_search.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2024 Mauro Romito
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{fmt::Debug, sync::Arc};
+
+use eyeball_im::VectorDiff;
+use futures_util::StreamExt;
+use matrix_sdk::room_directory_search::RoomDirectorySearch as SdkRoomDirectorySearch;
+use tokio::sync::RwLock;
+
+use super::RUNTIME;
+use crate::{error::ClientError, task_handle::TaskHandle};
+
+#[derive(uniffi::Enum)]
+pub enum PublicRoomJoinRule {
+    Public,
+    Knock,
+}
+
+impl PublicRoomJoinRule {
+    fn convert(value: ruma::directory::PublicRoomJoinRule) -> Option<Self> {
```

Comment:
Prefer [the `From` trait](https://doc.rust-lang.org/std/convert/trait.From.html):

```suggestion
impl From<ruma::directory::PublicRoomJoinRule> for PublicRoomJoinRule {
    fn from(value: ruma::directory::PublicRoomJoinRule) -> Self {
```

If it can fails, use [the `TryFrom` trait](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).